core/services/gateway/network/wsconnection.go

```diff
@@ -0,0 +1,104 @@
+package network
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/gorilla/websocket"
+)
+
+// ConnectionWrapper is a websocket connection abstraction that supports re-connects.
+// I/O is separated from connection management:
+//   - component doing I/O provides read and write channels that never change
+//   - component managing connections can listen to connection-closed events and call Restart()
+//
+// The Wrapper can be used by a server expecting long-lived connections from a given client,
+// as well as a client maintaining such long-lived connection with a given server.
+//
+// The concept of "pumps" borrowed from https://github.com/smartcontractkit/wsrpc
+type WSConnectionWrapper struct {
+	conn   *websocket.Conn
+	connMu sync.Mutex
```

Comment:
could these be simplified to an atomic.Pointer? https://pkg.go.dev/sync/atomic#Pointer
```suggestion
	conn   atomic.Pointer[websocket.Conn]
```
IIUC Swap() would work in Restart()